src/code/fault_n64.c

```diff
@@ -0,0 +1,838 @@
+                        x = 7;
+                        y = 1;
+                    } else {
+                        x = 0;
+                    }
+                    break;
+
+                case 7:
+                    if (btnPress == BTN_CRIGHT) {
+                        x = 8;
+                    } else {
+                        x = 0;
+                    }
+                    break;
+
+                case 8:
+                    if (btnPress == BTN_DRIGHT) {
+                        x = 9;
+                        y = 1;
+                    } else {
+                        x = 0;
+                    }
+                    break;
+
+                case 9:
+                    if (btnPress == (BTN_A | BTN_B)) {
+                        x = 10;
+                    } else if (btnPress == BTN_A) {
+                        x = 91;
+                    } else if (btnPress == BTN_B) {
+                        x = 92;
+                    } else {
+                        x = 0;
+                    }
+                    break;
+
+                case 91:
+                    if (btnPress == BTN_B) {
+                        x = 10;
+                    } else {
+                        x = 0;
+                    }
+                    break;
+
+                case 92:
+                    if (btnPress == BTN_A) {
+                        x = 10;
+                    } else {
+                        x = 0;
+                    }
+                    break;
+
+                case 10:
+                    if ((btnCur == (BTN_A | BTN_B | BTN_START)) && (btnPress == BTN_START)) {
+                        f32 comboTimeSeconds = OS_CYCLES_TO_USEC(osGetTime() - comboStartTime) / 1000000.0f;
+
+                        // Input time %f seconds
+                        osSyncPrintf("入力時間 %f 秒\n", comboTimeSeconds);
+                        if (comboTimeSeconds <= 50.0f) {
+                            x = 11;
+                        } else {
+                            x = 0;
+                        }
+                    } else {
+                        x = 0;
+                    }
+                    break;
+
+                default:
+                    break;
+            }
+        }
+    }
+}
+
+extern s32 D_80105A90_unknown; // Arena_failcnt
+
+void func_800AF0E0(void) {
+    s32 i;
+
+    Fault_DrawRecBlack(22, 16, 276, 208);
+    Fault_DrawText(40, 30, "SegmentBaseAddress");
+
+    for (i = 0; i < ARRAY_COUNT(gSegments); i++) {
+        Fault_DrawText(40, 40 + 8 * i, "%2d:%08x", i, gSegments[i]);
+    }
+    Fault_DrawText(40, 180, "Arena_failcnt = %d", D_80105A90_unknown);
+}
+
+void Fault_DrawMemDumpContents(const char* title, void* memory, u32 arg2) {
+    s32 x;
+    s32 y;
+    u32* ptr = (u32*)((uintptr_t)memory & ~3);
+
+    Fault_DrawRecBlack(22, 16, 276, 208);
+    Fault_DrawText(36, 18, "%s %08x", title != NULL ? title : "PrintDump", ptr);
+
+    if ((ptr >= (u32*)0x80000000) && (ptr <= (u32*)0x80400000)) {
```

Comment:
```suggestion
    if (((uintptr_t)ptr >= PHYS_TO_K0(0x000000)) && ((uintptr_t)ptr <= PHYS_TO_K0(0x400000))) {
```

Same for the rest of comparisons like this one